ios/observers/KeyboardMovementObserver.swift

```diff
@@ -263,20 +268,47 @@ public class KeyboardMovementObserver: NSObject {
     displayLink = nil
   }
 
-  @objc func updateKeyboardFrame() {
+  func initializeAnimation(fromValue: Double, toValue: Double) {
+    let positionAnimation = keyboardView?.layer.presentation()?.animation(forKey: "position") as? CASpringAnimation
+    guard let keyboardAnimation = positionAnimation else {
+      return
+    }
+    animation = SpringAnimation(animation: keyboardAnimation, fromValue: fromValue, toValue: toValue)
+  }
+
+  @objc func updateKeyboardFrame(link: CADisplayLink) {
     if keyboardView == nil {
       return
     }
 
     let keyboardFrameY = keyboardView?.layer.presentation()?.frame.origin.y ?? 0
     let keyboardWindowH = keyboardView?.window?.bounds.size.height ?? 0
-    let keyboardPosition = keyboardWindowH - keyboardFrameY
+    var keyboardPosition = keyboardWindowH - keyboardFrameY
 
     if keyboardPosition == prevKeyboardPosition || keyboardFrameY == 0 {
       return
     }
 
+    if let animation = animation {
+      let beginTime = animation.startTime
+      let baseDuration = link.targetTimestamp - beginTime
+
+      #if targetEnvironment(simulator)
+        let correctedDuration = baseDuration - UIUtils.nextFrame * 0.6
+      #else
+        let correctedDuration = baseDuration + UIUtils.nextFrame
+      #endif
+
+      let duration = correctedDuration
+      let position = CGFloat(animation.valueAt(time: duration))
+      // handles a case when final frame has final destination (i. e. 0 or 291)
+      // but CASpringAnimation can never get to this final destination
+      let race: (CGFloat, CGFloat) -> CGFloat = animation.isIncreasing ? max : min
+      keyboardPosition = race(position, keyboardPosition)
+    }
+
```

Comment:
Since we have condition above:

```swift
if keyboardPosition == prevKeyboardPosition
```

and here we assign potentially modified property- will not it cause problems? Do we need to remove the condition from above?